### ISSUE TITLE
Release 2.0 Sign-off List - Gathering Requirments

### ISSUE THREAD

**Issue by DanVanAtta** [triplea-game/triplea#6048]
Attempting to build a final sign-off list for 2.0 prerelease testing. Opening this issue to present an initial list and gather input for additional items that should be tested. Please add comments with additional items that should be included, the desired outcome is to have a mostly complete list of items that we should verify in the 2.0 prerelease.

-------------

### Lobby (L):

(L1) Chat:
L1.A) Chat messages are sent

(L2) Player listing:
L2.A) Moderators have (mod) tag
L2.B) Player listing right click menu shows a slap option
L3.C) Moderators only, shows a (working) disconnect and remove option in a right click menu
L4.D) Right clicking your own name does nothing

(L3) Slap:
L3.A) Slap messages are sent
L3.B) Slap message recipient sees "you have been slapped by X"
L3.C) Slap sender sees "you slapped X" 
L3.D) Everyone else sees "X slapped Y"

(L4) Banning:
L4.A) Banned players are removed from chat
L4.B) Banned players are removed from private host games
L4.C) Banned players are removed from bot games
L4.D) Games hosted by banned players are removed
L4.E) Banned players are unable to join lobby
L4.F) Banned players are unable to directly join hosted games
L4.G) Banned players are unable to directly join bot games
L4.H) On-ban, when a player is in lobby, a message is displayed to all users that a player was banned for violating lobby rules

(L5) Unbanning:
L5.A) Unbanning works, after unban player is able to join lobby and games

(L6) Username blocking:
L6.A) Username banning works
L6.B) What happens when we ban the username of a registered user?
L6.C) Users in lobby with a banned username are booted
L6.D) Users joining with a blocked name are unable to join with that name

(L7) Joining:
L7.A) registered users can join lobby multipe times and should see the same name used
L7.B) All players, including the joining player, should see "X has joined"
L7.C) joining with the same name that is already in lobby leads to a 'user has already joined' error message

(L8) Status Update
L8.A) Players can update their status, visible to everyone

(L9) Game Status
L9.A) Hosting a game posts game
L9.B) Changing game comments, player count changes, map changes, etc.. are reflected in lobby game status
L9.C) Launching game, stopping game, reflects properly in lobby as 'waiting', 'running', etc..
L9.D) Quitting a game delists the game from lobby

(L10) Player leaving lobby
L10.A) After leaving lobby a players name is removed from player list
L10.B) All players see a 'player has left' notification

(L11) Player disconnect
L11.A) Disconnecting a player removes them from lobby
L11.B) A "X was disconnected from lobby by moderator" message is shown to all users

(L12) Hosted Game Disconnect
L12.A) Moderators can disconnect hosted games from lobby (note, the game will auto-reconnect if it is still live)

(L13) Game Shutdown
L13.A) Moderators can issue remote shutdown to hosted games
L13.B) Hosted games that are remotely shutdown are removed from lobby (and they should not reconnect)

(L14) Auto-lobby reconnect
L14.A) Hosted games that are connected to lobby should show a lobby disconnect message if lobby is turned off
L14.B) HOsted games should automatically reconnect to a lobby that has been turned off and then turned back on

L15) Moderator Toolbox
L15.A) Only moderators can access toolbox
L15.B) Moderator actions show up in the toolbox log
L15.C) Moderators can ban and unban from the toolbox 
L15.D) Moderators can view, add and remove banned names from the toolbox
L15.E) Lobby access log is viewable from the toolbox (new player joins, new entry is added)
L15.F) Admins can promote users to moderator
L15.G) Admins can demote moderators to users
L15.H) Admins can promote moderators to admin

L16) Joining Lobby
L16A) Anonymous users cannot join if an existing user with the same name is already in lobby
  - The "(1)" suffix behavior is removed
L16B) Registered users get multiple windows if they are already logged in
  - There is no 'player has joined' notification except for the first time when a registered user joins multiple times

-------

### Gameplay (G)

G1) Sub combat rules (TODO: expound on this)
G2) Rocket rules (TODO: expound on this)
G3) Placement Undo
G3.A) Undo placement in 1941 with chinese placing without factories
G3.B) Undo standard placement on factories
G4) Battle Calc, test as thoroughly as possible, multiple windows, concurrently, different maps

 ### Bot/Game Hosting
- Verify IP address can be configured to an arbitrary value

**Comment by prastle** [comment]
After testing with Wirkey we have a few issues but I definitely believe it is close to a release.  I will attempt to click off all working functions now.
L14 works (and works WELL! gd job!) cant test rest since bot messed 

All L 15 work Except something that is missing in list.  If a banned name is added it shows up!  But it is not removed from the lobby nor banned from rejoining.   also It appears the banned names are broken but I will test more.  I will add to this list in a bit.

A mod or admin should not be able to ban a mod or admin they can

The signatory identifier of a multi use player is missing nor indicated
 more testing to come but as a general statement WELL DONE SIRS!

**Comment by prastle** [comment]
Players info is missing but I guess can be found in toolbox

**Comment by DanVanAtta** [comment]
> Players info is missing but I guess can be found in toolbox.

Yeah, it was questionable how much it was needed when the info is in toolbox. I realize it's pretty handy, though it aligns with an older process where we don't have a full information history.

> Slaps spam the lobby and all can see it. They never used to only the slapee and slaper could see it.

I think they are rate limited. The 'spam' is intentional so you can see if someone is slap spamming. Some players will slap any new player that joins, which is annoying.

> It would be preferred if only mods and player could see results of their actions please remove the general comment "prastle has been bad and removed from the lobby"

Why the preference? Players leaving after boot or ban sometimes looks like the player just left on their own accord. The public messaging is to tell everyone that the mod hammer was dropped.

**Comment by DanVanAtta** [comment]
Added a new item:

> Bot/Game Hosting
    Verify IP address can be configured to an arbitrary value

@RoiEXLab I'm not sure how to do that. Could you confirm if there is documentation on how that can be configured? Second, would you be able to verify the functionality works as expected?

**Comment by RoiEXLab** [comment]
@DanVanAtta I just checked and its doesn't work as expected. I'm not sure if I ever created an "official" guide, but here's the basic procedure:

1. Go to https://ngrok.com/download and follow the instructions steps 1 to 3 (you'll need to create a free account for this)
2. In the folder you extracted your executable to, run `ngrok tcp 4000`
3. In the run configuration for headless server add `-PcustomPort=17579 -PcustomHost=0.tcp.ngrok.io` (where you'll obviously replace domain name and port with the ones you were assigned by the server as shown in the client you just launched) and set the lobby parameter to `-Ptriplea.lobby.uri=https://prerelease.triplea-game.org`
4. Run the Headless client. The check tells you that your machine is not reachable from the internet and stops itself.

Here's what made me curious: Is it really not reachable, or is it just the check that verifies the wrong thing.
So I commented out the connectivity check and everything worked as expeczed, I could join my own game I just opened a port for without any port forwarding enabled (I double checked).

TL;DR What needs fixing is the connectivity check that checks the wrong thing. The rest is working fine.

**Comment by DanVanAtta** [comment]
Thanks for checking that @RoiEXLab . Would you mind following up to fix that? Second, we could use the above listed probably on the user guide with a forum post for questions that introduces the feature and links to the user guide. Probably we also need the parameters documented in the run_bot script.

**Comment by DanVanAtta** [comment]
Another thought as well, we probably should get an integration test (or some sort of test) put in place to make sure the behavior continues to work, most http functionality is tested (which was really useful, I did manage to break something lately and testing caught it)

**Comment by RoiEXLab** [comment]
@DanVanAtta I'll try to get a fix done. I might need some help in the process

**Comment by DanVanAtta** [comment]
Thanks @RoiEXLab , let me know how I can be of help if any.
